libraries/ms-drivers/src/stm32f0xx/mcp3427_adc.c

```diff
@@ -0,0 +1,180 @@
+
+  Mcp3427Channel current_channel =
+      (storage->config & (1 << MCP3427_CH_SEL_OFFSET)) >> MCP3427_CH_SEL_OFFSET;
+
+  storage->sensor_data[current_channel] = sensor_data;
+
+  if (current_channel == MCP3427_CHANNEL_2 && storage->callback != NULL) {
+    // We have all the data ready.
+    storage->callback((int16_t)storage->sensor_data[0], (int16_t)storage->sensor_data[1],
+                      storage->context);
+  }
+
+  event_raise(storage->data_trigger_event, prv_get_chip_identifier(storage));
+}
+
+// Trigger data read. Schedule a data ready event to be raised.
+static void prv_channel_trigger(struct Fsm *fsm, const Event *e, void *context) {
+  Mcp3427Storage *storage = (Mcp3427Storage *)context;
+  // We want to trigger a read. So we set the ready bit.
+  // If operating in continuous conversion mode, setting this bit has no effect.
+  storage->config |= MCP3427_RDY_MASK;
+  // Setting the current channel we want to read from. We just flip it from the previous read.
+  storage->config ^= (1 << MCP3427_CH_SEL_OFFSET);
+
+  i2c_write(storage->port, storage->addr, &storage->config, MCP3427_NUM_CONFIG_BYTES);
+  soft_timer_start_millis(MCP3427_MAX_CONV_TIME_MS, prv_raise_ready, storage, NULL);
+}
+
+// Lookup table for selected address. See manual table 5-3.
+static uint8_t s_addr_lookup[NUM_MCP3427_PIN_STATES][NUM_MCP3427_PIN_STATES] = {
+  { 0x0, 0x1, 0x2 },
+  { 0x3, 0x0, 0x7 },
+  { 0x4, 0x5, 0x6 },
+};
+
+StatusCode mcp3427_init(Mcp3427Storage *storage, Mcp3427Settings *settings) {
+  if (storage == NULL || settings == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  storage->data_ready_event = settings->adc_data_ready_event;
+  storage->data_trigger_event = settings->adc_data_trigger_event;
+
+  fsm_state_init(channel_1_trigger, prv_channel_trigger);
+  fsm_state_init(channel_1_readback, prv_channel_ready);
+  fsm_state_init(channel_2_trigger, prv_channel_trigger);
+  fsm_state_init(channel_2_readback, prv_channel_ready);
+  storage->port = settings->port;
+  storage->addr =
+      s_addr_lookup[settings->addr_pin_0][settings->addr_pin_1] | (MCP3427_DEVICE_CODE << 3);
+
+  // Writing configuration to the chip (see section 5.3.3 of manual).
+  // Note: Here, channel gets defaulted to 0.
+  uint8_t config = 0;
+  config |= (settings->conversion_mode << MCP3427_CONVERSION_MODE_OFFSET);
+  config |= (settings->sample_rate << MCP3427_SAMPLE_RATE_OFFSET);
+  config |= (settings->amplifier_gain << MCP3427_GAIN_SEL_OFFSET);
+  storage->config = config;
+
+  StatusCode status = i2c_write(storage->port, storage->addr, &config, MCP3427_NUM_CONFIG_BYTES);
+  // start the state machine ready to transition to channel_1_trigger
+  fsm_init(&storage->fsm, MCP3427_FSM_NAME, &channel_2_readback, storage);
+  return status;
+}
+
+StatusCode mcp3427_register_callback(Mcp3427Storage *storage, Mcp3427Callback callback,
+                                     void *context) {
+  if (storage == NULL || callback == NULL) {
```

Comment:
I think you can do !(storage && callback) but what you have is fine, this isn't really a big deal

Reply:
This appears in other parts too, so storage == NULL can become !storage or smthing. As I said this isn't really a big deal, up to you

Reply:
Lol, people gave the opposite comments when I did `!storage` last time. `== NULL` is probably slightly clearer.